kani-compiler/src/kani_middle/transform/check_uninit/ty_layout.rs

```diff
@@ -0,0 +1,275 @@
+            match ty.kind().rigid().expect(&format!("unexpected type: {ty:?}")) {
+                RigidTy::Adt(def, args) => {
+                    match def.kind() {
+                        AdtKind::Enum => {
+                            // Support basic enumeration forms
+                            let ty_variants = def.variants();
+                            match layout.variants {
+                                VariantsShape::Single { index } => {
+                                    // Only one variant is reachable. This behaves like a struct.
+                                    let fields = ty_variants[index.to_index()].fields();
+                                    let mut fields_validity = vec![];
+                                    for idx in layout.fields.fields_by_offset_order() {
+                                        let field_offset = offsets[idx].bytes();
+                                        let field_ty = fields[idx].ty_with_args(&args);
+                                        fields_validity.append(&mut data_bytes_for_ty(
+                                            machine_info,
+                                            field_ty,
+                                            field_offset + current_offset,
+                                        )?);
+                                    }
+                                    Ok(fields_validity)
+                                }
+                                VariantsShape::Multiple {
+                                    tag_encoding: TagEncoding::Niche { .. },
+                                    ..
+                                } => {
+                                    Err(format!("Unsupported Enum `{}` check", def.trimmed_name()))?
+                                }
+                                VariantsShape::Multiple { variants, .. } => {
+                                    let enum_validity = ty_size();
+                                    let mut fields_validity = vec![];
+                                    for (index, variant) in variants.iter().enumerate() {
+                                        let fields = ty_variants[index].fields();
+                                        for field_idx in variant.fields.fields_by_offset_order() {
+                                            let field_offset = offsets[field_idx].bytes();
+                                            let field_ty = fields[field_idx].ty_with_args(&args);
+                                            fields_validity.append(&mut data_bytes_for_ty(
+                                                machine_info,
+                                                field_ty,
+                                                field_offset + current_offset,
+                                            )?);
+                                        }
+                                    }
+                                    if fields_validity.is_empty() {
```

Comment:
I think in your case you want to check if their padding byte locations are the same. I think you also want to add the data where the variant lives to make sure that is initialized.

Reply:
Added a comment addressing that